File: scripts/postgresql-oracle.ts
import {backend} from '../native/postgresql/runtime';
import {importPostgresqlSql,exportPostgresqlSql,proposePostgresqlNodeEdit,readDocument,writeDocument} from '../src';
const cases=[];
for(const file of ['domain.sql','advanced.sql']){
 const source=await Bun.file('fixtures/postgresql/'+file).text();const doc=await importPostgresqlSql(source,backend,{id:file});
 const sql=await exportPostgresqlSql(readDocument(writeDocument(doc,'yaml'),'yaml'),backend);
 cases.push({name:file,source,sql});
}
const source='CREATE TABLE original (id integer);';const doc=await importPostgresqlSql(source,backend,{id:'edit'});const edit=proposePostgresqlNodeEdit(doc,'/stmts/0/stmt/CreateStmt/relation/relname','"changed"');
await Bun.write('fixtures/postgresql/oracle-input.json',JSON.stringify({backend:backend.identity,parserVersion:170004,cases,edit:{source,sql:await exportPostgresqlSql(edit.document,backend)}},null,2)+'\n');
const artifacts=[];for(const path of ['package.json','wasm/libpg-query.wasm','wasm/libpg-query.js','proto.js'])artifacts.push({path,sha256:new Bun.CryptoHasher('sha256').update(await Bun.file('node_modules/@libpg-query/parser/'+path).arrayBuffer()).digest('hex')});
await Bun.write('native/postgresql/runtime-manifest.json',JSON.stringify({package:'@libpg-query/parser',version:'17.6.10',observedPostgresqlVersion:170004,artifacts,adapter:{strategy:'Descriptor-directed nested message construction; Boolean.create bypasses the defective Boolean.fromObject; no global codec patch',sourceSha256:new Bun.CryptoHasher('sha256').update(await Bun.file('native/postgresql/runtime.ts').arrayBuffer()).digest('hex'),schemaSha256:new Bun.CryptoHasher('sha256').update(await Bun.file('spec/extensions/postgresql/native-ast.schema.json').arrayBuffer()).digest('hex')},scope:'Pinned package artifacts and observed raw parser version, not source-build reproducibility or PostgreSQL catalog compatibility'},null,2)+'\n');
console.log('PostgreSQL native oracle inputs generated');

File: scripts/postgresql-oracle.py
import json
from pathlib import Path
import pglast
from pglast.parser import parse_sql_json
inputs=json.loads(Path('fixtures/postgresql/oracle-input.json').read_text())
def tree(sql):
 value=json.loads(parse_sql_json(sql))
 def clean(x):
  if isinstance(x,list):return [clean(v) for v in x]
  if isinstance(x,dict):return {k:clean(v) for k,v in x.items() if k not in ['location','stmt_location','stmt_len']}
  return x
 return clean(value)
cases=[]
for row in inputs['cases']:
 a,b=tree(row['source']),tree(row['sql']);assert a==b,row['name']
 cases.append({'name':row['name'],'statements':len(a['stmts']),'nativeAstEqualExcludingOffsets':True})
original=tree(inputs['edit']['source']);edited=tree(inputs['edit']['sql']);assert original!=edited
original['stmts'][0]['stmt']['CreateStmt']['relation']['relname']='changed';assert original==edited
try:parse_sql_json('CREATE TABLE')
except Exception:invalid_rejected=True
else:raise AssertionError('Native parser accepted invalid SQL')
report={'oracle':'pglast '+pglast.__version__,'postgresqlVersion':pglast.get_postgresql_version(),'wasmParserVersion':inputs['parserVersion'],'cases':cases,'editChangesOnlySelectedName':True,'invalidSqlRejected':invalid_rejected,'scope':'Independent native C wrapper parsing original/re-exported SQL. Native PostgreSQL 17.7 versus WASM 17.4 are explicitly distinct. No catalog resolution, DDL execution or cross-system semantic equivalence claim.'}
Path('fixtures/postgresql/oracle-results.json').write_text(json.dumps(report,indent=2)+'\n')
print('PostgreSQL: two native schema round trips, isolated edit and invalid SQL rejection verified; native versions retained')
